File: csharp/1670.design-front-middle-back-queue.cs
/*
 * @lc app=leetcode id=1670 lang=csharp
 *
 * [1670] Design Front Middle Back Queue
 */

using System.Collections.Generic;

// @lc code=start
public class FrontMiddleBackQueue {

    private LinkedList<int> Left, Right;

    public FrontMiddleBackQueue() {
        Left = new LinkedList<int>();
        Right = new LinkedList<int>();
    }

    private void Balance() {
        while (Left.Count > Right.Count) {
            Right.AddFirst(Left.Last.Value);
            Left.RemoveLast();
        }
        while (Right.Count > Left.Count + 1) {
            Left.AddLast(Right.First.Value);
            Right.RemoveFirst();
        }
    }
    
    public void PushFront(int val) {
        Left.AddFirst(val);
        Balance();
    }
    
    public void PushMiddle(int val) {
        Left.AddLast(val);
        Balance();
    }
    
    public void PushBack(int val) {
        Right.AddLast(val);
        Balance();
    }
    
    public int PopFront() {
        int ret = -1;
        if (Left.Count == 0) {
            if (Right.Count == 0) {
                return -1;
            }
            ret = Right.First.Value;
            Right.RemoveFirst();
        } else {
            ret = Left.First.Value;
            Left.RemoveFirst();
        }
        Balance();
        return ret;
    }
    
    public int PopMiddle() {
        int ret = -1;
        if (Left.Count == 0 && Right.Count == 0) {
            return -1;
        }
        if (Left.Count == Right.Count) {
            ret = Left.Last.Value;
            Left.RemoveLast();
        } else {
            ret = Right.First.Value;
            Right.RemoveFirst();
        }
        Balance();
        return ret;
    }
    
    public int PopBack() {
        if (Right.Count == 0) {
            return -1;
        }
        int ret = Right.Last.Value;
        Right.RemoveLast();
        Balance();
        return ret;
    }
}

/**
 * Your FrontMiddleBackQueue object will be instantiated and called as such:
 * FrontMiddleBackQueue obj = new FrontMiddleBackQueue();
 * obj.PushFront(val);
 * obj.PushMiddle(val);
 * obj.PushBack(val);
 * int param_4 = obj.PopFront();
 * int param_5 = obj.PopMiddle();
 * int param_6 = obj.PopBack();
 */
// @lc code=end

File: java/1670.design-front-middle-back-queue.java
/*
 * @lc app=leetcode id=1670 lang=java
 *
 * [1670] Design Front Middle Back Queue
 */

import java.util.ArrayDeque;
import java.util.Deque;

// @lc code=start

class FrontMiddleBackQueue {

    private final Deque<Integer> l, r;

    public FrontMiddleBackQueue() {
        l = new ArrayDeque<>();
        r = new ArrayDeque<>();
    }
    
    public void pushFront(int val) {
        l.addFirst(val);
        balance();
    }
    
    public void pushMiddle(int val) {
        r.addFirst(val);
        balance();
    }
    
    public void pushBack(int val) {
        r.addLast(val);
        balance();
    }
    
    public int popFront() {
        int ret = l.isEmpty() ? (r.isEmpty() ? -1 : r.removeFirst()) : l.removeFirst();
        balance();
        return ret;
    }
    
    public int popMiddle() {
        int ret = r.isEmpty() ? -1 : l.size() == r.size() ? l.removeLast() : r.removeFirst();
        balance();
        return ret;
    }
    
    public int popBack() {
        int ret = r.isEmpty() ? -1 : r.removeLast();
        balance();
        return ret;
    }

    private void balance() {
        while (l.size() > r.size()) {
            r.addFirst(l.removeLast());
        }
        while (r.size() > l.size() + 1) {
            l.addLast(r.removeFirst());
        }
    }
}

/**
 * Your FrontMiddleBackQueue object will be instantiated and called as such:
 * FrontMiddleBackQueue obj = new FrontMiddleBackQueue();
 * obj.pushFront(val);
 * obj.pushMiddle(val);
 * obj.pushBack(val);
 * int param_4 = obj.popFront();
 * int param_5 = obj.popMiddle();
 * int param_6 = obj.popBack();
 */
// @lc code=end

